### PR TITLE
Add validation reason and distinguish between live and batch validation

### PR DESCRIPTION
**Pull request by planger**
https://github.com/eclipse-glsp/glsp/issues/980

For review, please see [the comment on which changes belong together](https://github.com/eclipse-glsp/glsp/issues/980#issuecomment-1516867263).

### CODE REVIEW

File: plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java [hunk 1]
```diff
@@ -23,14 +25,73 @@
...
     * @param elements The list of {@link GModelElement} to validate.
+    * @param reason   The reason for a validation request, such as "batch" or "live" validation.
     * @return A list of {@link Marker}s for the validated {@link GModelElement}s.
     */
-   List<Marker> validate(GModelElement... elements);
+   default List<Marker> validate(final List<GModelElement> elements, final String reason) {
+      List<Marker> markers = new ArrayList<>();
+
+      for (GModelElement element : elements) {
+         if (MarkersReason.LIVE.equals(reason)) {
```
Comment (tortmayr) [inline plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java:50]: What is the assumption here regarding the reason property? Is it a fixed string that can only either be `batch` or `live`? Or can adopters also introduce custom reasons? If so we should probably explicitly check for the `batch` reason as well and add a fallback case for unsupported reasons.

File: plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java [hunk 2]
```diff
@@ -23,14 +25,73 @@
...
+            markers.addAll(doLiveValidation(element));
+         } else {
+            markers.addAll(doBatchValidation(element));
+         }
+         if (element.getChildren() != null) {
```
Comment (tortmayr) [inline plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java:55]: I think this is a unnecessary null-check. List properties of EObjects are always defined and should never be null. We should rather check if the children list is not empty.

File: plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/RequestMarkersHandler.java [hunk 3]
```diff
@@ -43,27 +42,24 @@ public class RequestMarkersHandler extends AbstractActionHandler<RequestMarkersA
...
-      List<Marker> markers = new ArrayList<>();
-      GModelIndex currentModelIndex = modelState.getIndex();
-      for (String elementID : elementsIDs) {
-         Optional<GModelElement> modelElement = currentModelIndex.get(elementID);
-         if (modelElement.isPresent()) {
-            markers.addAll(validator.get().validate(modelElement.get()));
-         }
 
-      }
+      String reason = action.getReason() != null ? action.getReason() : MarkersReason.BATCH;
```
Comment (tortmayr) [inline plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/RequestMarkersHandler.java:57]: See comment above. If the default is already properly set in the action class we don't need to handle it explicitly here.

**Review by tortmayr — CHANGES_REQUESTED** [review]
Thanks Philip, support for live validation is definitely a nice feature addition.
I did not have time to test the changes yet and only did a preliminary code review for now.

File: plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java (hunk 1 above)
Comment (planger) [reply plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java:50]: Right, the assumption is that we use batch or live by default, but adopters may add custom reasons, e.g. some specific analysis and use custom strings here. My thinking was that they then would have to overwrite the `validate()` method. But you are right, it'd be more convenient to handle the batch case explicitly and have a `default doValidationForCustomReason()` that needs to be overwritten in those cases.

**Comment by planger** [comment]
Thank you for the excellent review! I've applied your suggestions with b8a020f.

**Review by tortmayr — APPROVED** [review]
Thanks for the fast update! LGTM 👍🏼

**Pull request merged**